helix-rest/src/test/java/org/apache/helix/rest/server/TestClusterAccessor.java

```diff
@@ -259,7 +259,15 @@ public void testUpdateConfigFields() throws IOException {
     record.getSimpleFields().clear();
     record.setSimpleField(key, value + "--updated");
 
-    key = record.getListFields().keySet().iterator().next();
+    // skip INSTANCE_CAPACITY_KEYS which can be added by waged rebalancer in global run
```

Comment:
I think an easier way is to reconfigure the dependencies so that all the tests in this class run sequentially.

The shortage of current solutions is that it mixed up the logic between different tests. And once we have more test cases, it may break again.

Reply:
Not sure I got what you mean here. Current testng.xml already let the test run sequentially. 

```
<suite name="Suite" parallel="false">
  <test name="Test" preserve-order="true">
    <packages>
      <package name="org.apache.helix.rest.*"/>
    </packages>
  </test>
</suite>
```

Reply:
Can you draw some code to illustrate the idea?

Reply:
As discussed offline, we agreed that by enforcing the order of test cases, we shall be able to avoid this failure easily without touching the test logic.

Reply:
@jiajunwang, I think your theory is not right. The following is dependency trail. 

>testEnableWagedRebalanceForAllResources ->
testActivateSuperCluster ->
testGetStateModelDef ->
testEnableDisableMaintenanceModeWithCustomFields ->
testGetMaintenanceHistory ->
testGetControllerLeadershipHistory ->
testEnableDisableMaintenanceMode ->
testGetClusterConfig ->
testEnableDisableCluster ->
testCreateDeleteCluster ->
testDeleteConfigFields ->
testUpdateConfigFields

Reply:
Let us just use the current approach as waged is newest one and would be here for a long time.  Further change may also break it.

Reply:
That was a guess, yeah. But the key is the current way is not the correct solution, IMHO.
If our conclusion was wrong, that means we still don't understand the root cause. I would say we need to clearly understand why the WAGED field was added without any related test running. Then we can fix it in the right way.

Do you have the github environment execute order so I can reproduce it?

Reply:
https://github.com/apache/helix/runs/1246249153?check_suite_focus=true

look at log in this link or any other test in rest.

Reply:
That said, this is not guess. See the analysis in #1336. Basically, the only code path is following:

>computeBestPossibleAssignment -->
globalRebalance -->
ResourceChangeDetector.updateSnapshots -->
ResourceChangeSnapshot(ResourceControllerDataProvider dataProvider,
boolean ignoreNonTopologyChange) -->
ClusterConfigTrimmer.getInstance().trimProperty(dataProvider.getClusterConfig()) -->
HelixPropertyTrimmer.doTrim -->
copyNonTrimmableInfo(originalZNRecord, trimmedZNRecord, getNonTrimmableFields(originalProperty),
false) -->
protected Map<FieldType, Set> getNonTrimmableFields(ClusterConfig property) {
return STATIC_TOPOLOGY_RELATED_FIELD_MAP;
} -->
private static final Map<FieldType, Set> STATIC_TOPOLOGY_RELATED_FIELD_MAP = ImmutableMap
.of(FieldType.SIMPLE_FIELD, ImmutableSet
.of(ClusterConfigProperty.TOPOLOGY.name(),
ClusterConfigProperty.FAULT_ZONE_TYPE.name(),
ClusterConfigProperty.TOPOLOGY_AWARE_ENABLED.name(),
ClusterConfigProperty.MAX_PARTITIONS_PER_INSTANCE.name()),
FieldType.LIST_FIELD, ImmutableSet
.of(ClusterConfigProperty.INSTANCE_CAPACITY_KEYS.name()),
FieldType.MAP_FIELD, ImmutableSet
.of(ClusterConfigProperty.DEFAULT_INSTANCE_CAPACITY_MAP.name(),
ClusterConfigProperty.DEFAULT_PARTITION_WEIGHT_MAP.name(),
ClusterConfigProperty.REBALANCE_PREFERENCE.name()));

note 

```
private void copyNonTrimmableInfo(ZNRecord originalZNRecord, ZNRecord trimmedZNRecord,
      Map<FieldType, Set<String>> nonTrimmableFields, boolean trimValue) {
    for (Map.Entry<FieldType, Set<String>> fieldEntry : nonTrimmableFields.entrySet()) {
      FieldType fieldType = fieldEntry.getKey();
      Set<String> fieldKeySet = fieldEntry.getValue();
      if (null == fieldKeySet || fieldKeySet.isEmpty()) {
        continue;
      }
      switch (fieldType) {
        case SIMPLE_FIELD:
          fieldKeySet.stream().forEach(fieldKey -> {
            if (originalZNRecord.getSimpleFields().containsKey(fieldKey)) {
              trimmedZNRecord.getSimpleFields().putIfAbsent(fieldKey,
                  trimValue ? null : originalZNRecord.getSimpleField(fieldKey));
            }
          });
          break;
        case LIST_FIELD:
          fieldKeySet.stream().forEach(fieldKey -> {
            if (originalZNRecord.getListFields().containsKey(fieldKey)) {
              trimmedZNRecord.getListFields().putIfAbsent(fieldKey,
                  trimValue ? **Collections.EMPTY_LIST** : originalZNRecord.getListField(fieldKey));
            }
          });
          break;

```
This would add an empty INSTANCE_CAPACITY_KEYS=[]

Reply:
I don't get it. computeBestPossibleAssignment shall not lead to any config change being written back to ZK. If so, then we need to fix this bug. Are you saying the trimmed result is written back to ZK?

Reply:
In this case, before `computeBestPossibleAssignment`, there is no key INSTANCE_CAPACITY_KEYS. After this call, it seems not such key, it would add INSTANCE_CAPACITY_KEYS=[].

It does not really trim anything, but still assume that INSTANCE_CAPACITY_KEYS is empty. Thus, it would explicitly write it out.

Reply:
1. This logic will only be executed when WAGED is used for at least one resource.
2. This logic update in-memory object only, so could you please identify the "write it out" part? I think what you meant is a ZK accessor write? If so, then it would be a bug, we shall fix it.